src/core/lib/transport/metadata_batch.h

```diff
@@ -185,6 +190,88 @@ class MetadataMap {
       GRPC_MUST_USE_RESULT;
   void MaybeUnlinkCallout(grpc_linked_mdelem* storage);
 
+  static void assert_valid_list(grpc_mdelem_list* list) {
+#ifndef NDEBUG
+    grpc_linked_mdelem* l;
+
+    GPR_ASSERT((list->head == nullptr) == (list->tail == nullptr));
+    if (!list->head) return;
+    GPR_ASSERT(list->head->prev == nullptr);
+    GPR_ASSERT(list->tail->next == nullptr);
+    GPR_ASSERT((list->head == list->tail) == (list->head->next == nullptr));
+
+    size_t verified_count = 0;
+    for (l = list->head; l; l = l->next) {
+      GPR_ASSERT(!GRPC_MDISNULL(l->md));
+      GPR_ASSERT((l->prev == nullptr) == (l == list->head));
+      GPR_ASSERT((l->next == nullptr) == (l == list->tail));
+      if (l->next) GPR_ASSERT(l->next->prev == l);
+      if (l->prev) GPR_ASSERT(l->prev->next == l);
+      verified_count++;
+    }
+    GPR_ASSERT(list->count == verified_count);
+#else
+    // Avoid unused-parameter warning for debug-only parameter
+    (void)list;
+#endif /* NDEBUG */
+  }
+
+  static grpc_error_handle GPR_ATTRIBUTE_NOINLINE
```

Comment:
Why can't this be inlined?

Reply:
Typically for cold paths near hot paths we'll extract the rarely executed code into a noinline function so that the commonly executed code around it can be better optimized by a compiler - it often prevents spilling registers to the stack for instance.

Reply:
(note that this is a code movement pr and I expect to delete most of this code within a month)

Reply:
Oh interesting, thanks for the info.